### PR TITLE
[FLINK-30799][runtime] Make SinkFunction support speculative execution for batch jobs

### PR DESCRIPTION
**Pull request by ifndef-SleePy**
## What is the purpose of the change

*This pull request introduces a way to enable speculative execution for `SinkFunction`*

## Brief change log

  - *Speculative execution would be enabled if a `SinkFunction` inherits `SupportsConcurrentExecutionAttempts`*

## Verifying this change

This change added tests:
  - *Add new test cases in `StreamingJobGraphGeneratorTest`*
  - *Add new test cases in `SpeculativeSchedulerITCase`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by ifndef-SleePy** [comment]
@zhuzhurk Please take a look when you have time.
By the way, the failed test case seems unstable. I think it's not caused by this PR.

File: flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java [hunk 1]
```diff
@@ -184,6 +185,23 @@ public void testSpeculativeSlowSink() throws Exception {
...
+    @Test
+    public void testNonSpeculativeSlowSinkFunction() throws Exception {
+        executeJob(this::setupNonSpeculativeSlowSinkFunction);
+        waitUntilJobArchived();
+
+        checkResults();
+    }
+
+    @Test
+    @Timeout(30)
```
Comment (zhuzhurk) [inline flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java:197]: `Timeout` should be avoided for Flink tests. See https://flink.apache.org/contributing/code-style-and-quality-common.html#avoid-timeouts-in-junit-tests.

**Review by zhuzhurk — COMMENTED** [review]
Thanks for creating this PR! @ifndef-SleePy 
I only have two minor comments.

**Comment by ifndef-SleePy** [comment]
@zhuzhurk Thanks for reviewing! I have addressed the comments.

**Review by zhuzhurk — APPROVED** [review]
LGTM.

**Comment by ifndef-SleePy** [comment]
Azure testing failed again, it's caused by https://issues.apache.org/jira/browse/FLINK-30727

**Comment by ifndef-SleePy** [comment]
Shall we re-trigger Azure testing or merge this pr directly? @zhuzhurk

**Comment by zhuzhurk** [comment]
Will merge it since the case is pretty unstable.

**Pull request closed without merging**